src/ILLink.Shared/TrimAnalysis/ValueHolder.cs

```diff
@@ -0,0 +1,61 @@
+﻿using System.Diagnostics;
+using MultiValue = ILLink.Shared.DataFlow.ValueSet<ILLink.Shared.DataFlow.SingleValue>;
+
+namespace ILLink.Shared.TrimAnalysis
+{
+	internal partial class ValueHolder : IValueHolder
+    {
+		public MultiValue Value { get; private set; }
+		internal ValueHolderKind Kind { get; init; }
+		internal ValueHolder (MultiValue value, ValueHolderKind kind)
+		{
+			Value = value;
+			Kind = kind;
+		}
+		internal partial class ValueHolderReference<TKey> : IValueHolder
```

Comment:
I don't really understand what is this type for.
I think I understand the idea behind the ValueHolder itself, but not really this reference type.
The linker already has a notion of a reference - that's basically what the ByRef field on stack slot is for. Why do we need a similar concept again? Unless you want to replace one with the other...